### PR TITLE
Added KroneckerDelta derivatives

### PR DESCRIPTION
**Pull request by nishnik**

### CODE REVIEW

File: symengine/tests/basic/test_functions.cpp [hunk 1]
```diff
@@ -2069,6 +2072,14 @@ TEST_CASE("Kronecker Delta: functions", "[functions]")
     r2 = one;
     REQUIRE(eq(*r1, *r2));
 
+    r1 = kronecker_delta(i, i)->diff(i);
```
Comment (isuruf) [inline symengine/tests/basic/test_functions.cpp]: This is not testing a `KroneckerDelta` instance

File: symengine/tests/basic/test_functions.cpp [hunk 2]
```diff
@@ -2069,6 +2072,14 @@ TEST_CASE("Kronecker Delta: functions", "[functions]")
...
+    r2 = zero;
+    REQUIRE(eq(*r1, *r2));
+
+    r1 = kronecker_delta(mul(i, i), j)->diff(i);
+    r2 = mul(i, mul(integer(2), Subs::create(Derivative::create(kronecker_delta(_x, j), {_x}), {{_x, mul(i, i)}})));
+    REQUIRE(eq(*r1, *r2));
```
Comment (isuruf) [inline symengine/tests/basic/test_functions.cpp]: Can you also add more tests for,
```
kronecker_delta(i, j)->diff(i);
kronecker_delta(i, j)->diff(j);
kronecker_delta(i, mul(j, j))->diff(j);
```

**Comment by nishnik** [comment]
ping @isuruf

File: symengine/tests/basic/test_functions.cpp [hunk 3]
```diff
@@ -2071,6 +2074,26 @@ TEST_CASE("Kronecker Delta: functions", "[functions]")
...
+    r1 = kronecker_delta(i, i)->diff(i);
+    r2 = zero;
+    REQUIRE(eq(*r1, *r2));
+
+    r1 = kronecker_delta(i, j)->diff(i);
+    r2 = Subs::create(Derivative::create(kronecker_delta(_x, j), {_x}), {{_x, i}});
+    REQUIRE(eq(*r1, *r2));
+
+    r1 = kronecker_delta(i, j)->diff(j);
+    r2 = Subs::create(Derivative::create(kronecker_delta(_x, i), {_x}), {{_x, j}});
```
Comment (isuruf) [inline symengine/tests/basic/test_functions.cpp]: This should be `Derivative::create(kronecker_delta(j, i), {j})`

**Comment by nishnik** [comment]
ping @isuruf

File: symengine/tests/basic/test_functions.cpp [hunk 4]
```diff
@@ -2071,6 +2076,31 @@ TEST_CASE("Kronecker Delta: functions", "[functions]")
...
+    r1 = kronecker_delta(i, j)->diff(j);
+    r2 = Derivative::create(kronecker_delta(i, j), {j});
+    REQUIRE(eq(*r1, *r2));
+
+    r1 = kronecker_delta(i, mul(j, j))->diff(j);
+    r2 = mul(j, mul(integer(2), Derivative::create(kronecker_delta(i, mul(j, j)), {j})));
+    REQUIRE(eq(*r1, *r2));
+
+    r1 = kronecker_delta(mul(i, i), j)->diff(i);
+    r2 = mul(i, mul(integer(2), Derivative::create(kronecker_delta(mul(i, i), j), {i})));
```
Comment (isuruf) [inline symengine/tests/basic/test_functions.cpp]: This is not correct. See SymPy output

**Comment by nishnik** [comment]
ping @isuruf

**Comment by isuruf** [comment]
Thanks

**Pull request merged**